[mlir][docs] Add C example for C-compatible wrapper for LLVM IR

ZenithalHourlyRate opened the pull request:
`TargetLLVMIR` documentation introduced the C-compatible wrapper function for a MLIR function and ways to generate it, but did not demonstrate the corresponding C function signature for them.

The C function signature is not obvious, in that
* `MemrefDescriptor` should be passed as _pointer_.
  + For example, MLIR function could return a new Descriptor, so pointer is a must.
  + Surprisingly, directly pass the struct, by C convention, is also a pointer so some function will work, but that is implicit and error-prone.
* for `@foo() -> memref<>`, the return type becomes the first argument in `_mlir_ciface_foo(%arg0: !llvm.ptr)`.
  + This is described in https://github.com/llvm/llvm-project/blob/f70ab7d909d6861c7eec5ab40679bde16ab826c6/mlir/lib/Conversion/FuncToLLVM/FuncToLLVM.cpp#L110-L167 Especially by code `size_t argOffset = resultStructType ? 1 : 0;` saying the actual argument starts at 1 when result is a struct (memref)

Users using the wrong signature will get incorrect results. LLVM discourse has some example of it
* https://discourse.llvm.org/t/how-to-compile-and-link-with-other-c-c-programs/4835/10
* https://discourse.llvm.org/t/segmentation-fault-on-memref-store/80286/3
* https://discourse.llvm.org/t/memref-store-storing-a-memref-load/80307

Cc @ftynse for relevent commit history. Cc @charitha22 and @Wheest from discourse post.

ZenithalHourlyRate committed:
[mlir][docs] Add C example for C-compatible wrapper for LLVM IR

ftynse commented on mlir/docs/TargetLLVMIR.md:
````diff
@@ -683,8 +683,18 @@ llvm.func @qux(%arg0: !llvm.ptr, %arg1: !llvm.ptr,
 
+
+```cpp
+// The C function implementation for the interface function
````
```suggestion
// The C function implementation for the interface function.
```

ftynse commented on mlir/docs/TargetLLVMIR.md:
````diff
@@ -719,8 +729,15 @@ llvm.func @_mlir_ciface_foo(%arg0: !llvm.ptr) {
 ```
 
+```cpp
+// The C function signature for the interface function
````
```suggestion
// The C function signature for the interface function.
```

ftynse commented on mlir/docs/TargetLLVMIR.md:
````diff
@@ -760,6 +778,14 @@ llvm.func @_mlir_ciface_foo(%arg0: !llvm.ptr, %arg1: !llvm.ptr) {
 ```
 
+```cpp
+// The C function signature for the interface function
````
```suggestion
// The C function signature for the interface function.
```

ftynse commented on mlir/docs/TargetLLVMIR.md:
```diff
@@ -744,6 +761,7 @@ llvm.func @foo(%arg0: !llvm.ptr, %arg1: !llvm.ptr, %arg2: i64,
 }
 
 // Interface function callable from C.
+// NOTE that the returned memref becomes the first argument
```
```suggestion
// NOTE: the returned memref becomes the first argument.
```

ftynse commented:
Thanks, documentation improvements are always appreciated! Let me know if you need help merging this.

ZenithalHourlyRate commented:
Addressed the review comment

> Let me know if you need help merging this.

I do not have commit access for now. I'd appreciate it if you could do me the favor.
